daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala

```diff
@@ -247,4 +245,75 @@ object TransactionBuilder {
   val EmptySubmitted: SubmittedTransaction = SubmittedTransaction(Empty)
   val EmptyCommitted: CommittedTransaction = CommittedTransaction(Empty)
 
+  def assignVersion[Cid](
+      v0: Value,
+      supportedVersions: VersionRange[TransactionVersion] = TransactionVersion.StableVersions
+  ): Either[String, TransactionVersion] = {
+    @tailrec
+    def go(
+        currentVersion: TransactionVersion,
+        values0: FrontStack[Value],
+    ): Either[String, TransactionVersion] = {
+      import LfValue._
+      if (currentVersion >= supportedVersions.max) {
+        Right(currentVersion)
+      } else {
+        values0 match {
+          case FrontStack() => Right(currentVersion)
+          case FrontStackCons(value, values) =>
+            value match {
+              // for things supported since version 1, we do not need to check
+              case ValueRecord(_, fs) => go(currentVersion, fs.map(v => v._2) ++: values)
+              case ValueVariant(_, _, arg) => go(currentVersion, arg +: values)
+              case ValueList(vs) => go(currentVersion, vs.toImmArray ++: values)
+              case ValueContractId(_) | ValueInt64(_) | ValueText(_) | ValueTimestamp(_) |
+                  ValueParty(_) | ValueBool(_) | ValueDate(_) | ValueUnit =>
+                go(currentVersion, values)
+              case ValueNumeric(_) =>
```

Comment:
Is there a reason why this is not part of the previous case?

Reply:
historical reason. I fix that.